enhancements/20240930-vxlan-enhancement.md

```diff
@@ -0,0 +1,197 @@
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.22
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.33
+bridge fdb append 02:22:22:22:22:22 dev vxlan42 dst 192.168.5.22
+bridge fdb append 02:33:33:33:33:33 dev vxlan42 dst 192.168.5.33
+
+on NODE_2:	add Mgmt NODE and NODE_3 in the forwarding table.
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.11
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.33
+bridge fdb append 02:11:11:11:11:11 dev vxlan42 dst 192.168.5.11
+bridge fdb append 02:33:33:33:33:33 dev vxlan42 dst 192.168.5.33
+
+on NODE_3:	add Mgmt NODE and NODE_2 
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.11
+bridge fdb append 00:00:00:00:00:00 dev vxlan42 dst 192.168.5.22
+bridge fdb append 02:11:11:11:11:11 dev vxlan42 dst 192.168.5.11
+bridge fdb append 02:22:22:22:22:22 dev vxlan42 dst 192.168.5.22
+```
+
+(5) The VxLAN interface on each NODE should play as the ARP proxy for VMs attached to the same VxLAN network.
+Then the packet destined to the VM have the ARP response from the corresponding VTEP and correctly forwarded.
+Adding the ARP entry for each VM for other NODEs which has no information for it. 
+```
+if the VM is on the Mgmt NODE, the following arp entry needs to be added on
+NODE_2/NODE_3:
+sudo ip neigh add 10.42.0.21 lladdr 02:11:11:11:11:11 dev vxlan0
+
+if the VM is on the NODE_2, then we need to add the ARP entry on
+Mgmt NODE/NODE_3:
+sudo ip neigh add 10.42.0.21 lladdr 02:22:22:22:22:22 dev vxlan0
+```
+
+(6) Use the VxLAN interface created in (3) to play as the router for all VMs attached to the same VxLAN network on the NODE.
+The `vxlan` interface can connect the external network via the uplink/bonding interface of the guest ClusterNetwork.
+Thus all VMs which route via the vxlan interface can connect to the external network.
+The `iptables` rules which do the masquerade should be added and the `default` route entry should also be configured on the attached VM.
+```
+if the VM resides on the Mgmt NODE:
+sysctl -w net.ipv4.ip_forward=1
+iptables -t nat -s 10.42.0.0/24 -A POSTROUTING -j MASQUERADE
+ip addr add 10.42.0.1 dev vxlan42	(or automatically assigned by IPAM/DHCP)
+
+on the VM:
+ip route add default dev enp2s0 via 10.42.0.1
+```
+
+### API changes
+
+## Design
+
+### Implementation Overview
+(1) Extend current network-controller to create the extra vxlan interface, manipulate FDB and ARP entires.
+
+(2) Use kube-virt to create VM, allocate the IP from the specified CIDR for the vxlan, and set up the default route correctly
+
+(3) network-controller need to keep tracking the VTEP IP/MAC mapping information for all NODEs. 
+
+### Test plan
+
+
+### Upgrade strategy
+
+## Note [optional]
+
+(1) to enable cross VxLAN network connection (different VNI) on the same guest Cluster Network, the BGP EVPN is required to route for each VxLAN group. Need to figure out a way for it.
```

Comment:
No, you don't need any BGP-EVPN to get cross vxlan network connection.
BGP-EVPN only helps you to configure route table correctly.
What you need is L3VNI. (Please google it, you don't need BGP-EVPN to do this, just because configuring manually is complex)
You still can configure those route table and configuration on your own.
Even you can write a K8s Controller to replace BGP-EVPN's role.